src/Hooks.ts

```diff
@@ -143,3 +145,75 @@ export function useFocusState() {
 
   return focusState;
 }
+
+type EffectCallback = (() => void) | (() => () => void);
+
+// Inspired by same hook from react-navigation v5
+// See https://github.com/react-navigation/hooks/issues/39#issuecomment-534694135
+export const useFocusEffect = (callback: EffectCallback) => {
+  const navigation = useNavigation();
+  const getCallback = useGetter(callback);
+
+  useEffect(() => {
+    let isFocused = false;
+    let cleanup: (() => void) | void;
+
+    if (navigation.isFocused()) {
+      cleanup = getCallback()();
+      isFocused = true;
+    }
+
+    const focusSubscription = navigation.addListener('willFocus', () => {
+      // If callback was already called for focus, avoid calling it again
+      // The focus event may also fire on intial render, so we guard against runing the effect twice
+      if (isFocused) {
+        return;
+      }
+
+      cleanup && cleanup();
+      cleanup = getCallback()();
+      isFocused = true;
+    });
+
+    const blurSubscription = navigation.addListener('willBlur', () => {
+      cleanup && cleanup();
+      cleanup = undefined;
+      isFocused = false;
+    });
+
+    return () => {
+      cleanup && cleanup();
+      focusSubscription.remove();
+      blurSubscription.remove();
+    };
+  }, [getCallback, navigation]);
+};
+
+export const useIsFocused = () => {
+  const navigation = useNavigation();
+  const [focused, setFocused] = useState(navigation.isFocused());
+
+  useEffect(() => {
+    const focusSubscription = navigation.addListener('willFocus', () =>
+      setFocused(true),
+    );
+    const blurSubscription = navigation.addListener('willBlur', () =>
+      setFocused(false),
+    );
+    return () => {
+      focusSubscription.remove();
+      blurSubscription.remove();
+    };
+  }, [setFocused]);
+
+  return focused;
+};
+
+export const useBackHandler = (backHandler: () => boolean) => {
+  useFocusEffect(() => {
```

Comment:
This can be simplified to:

```js
useFocusEffect(() => {
  const subscription = BackHandler.addEventListener('hardwareBackPress', backHandler);
  return subscription.remove();
})
```

Though I'm not convinced that this needs to be in here since it's little code.

It's also worth noting that this lib could be used in pure react projects without react-native-web, but adding an import from `react-native` is a breaking change. Though highly unlikely that anyone used it that way.

Reply:
ah that's right. 

But couldn't a web user pick a react-native-web implementation of BackHandler then, which is no-op? will check in RNW but anyway this means that we should publish a bundle for web that imports BackHandler from react-native-web (i think?). 

I'd rather add this in doc as a recipe

Reply:
Yeah, but I said "without react-native-web". React Navigation core can be used that way, but again I don't think anyone is using it without RNW or RN.

Reply:
Are you planning to add something similar in v5? If not I'll do like you and just add a doc recipe, so that user don't get confused when upgrading.

That seems pretty safe (RN is already a peerDependency and I guess we can expect to find it in the runtime), but let's keep things consistant between v4/v5

Reply:
We don't have any plans to add it in v5. If we were to add something like this, I would personally prefer something which also works for preventing going back from anywhere and not just the back button.

Reply:
That makes sense. I'm also doing this in my apps where I have a stack header and if a custom onBack callback is provided, I also wire the back button to it, and allow to return the backhandler boolean. This way if needed to prevent back navigation, the RN and hardware buttons remain sync.


Will remove this hook then